From 7e7b1c6b1bd3af168f3e27eb44c6749433f77c5c Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Sun, 25 Jan 2026 16:11:18 -0600
Subject: [PATCH 163/268] ntdll: Implement NtDelayExecution relative wait using
 QPC.

Some applications are very timing sensitive to the point that NTP adjustments throw them off. Surely the problem lies
somewhere else...

Arknights: Endfield requires the yield for some reason and it's there in the other code path as well. Therefore, we should probably yield.
---
 dlls/ntdll/unix/sync.c | 25 ++++++++++++++++++++-----
 1 file changed, 20 insertions(+), 5 deletions(-)

diff --git a/dlls/ntdll/unix/sync.c b/dlls/ntdll/unix/sync.c
index 2b824d45706..f8d0a95611a 100644
--- a/dlls/ntdll/unix/sync.c
+++ b/dlls/ntdll/unix/sync.c
@@ -2472,16 +2472,31 @@ NTSTATUS WINAPI NtDelayExecution( BOOLEAN alertable, const LARGE_INTEGER *timeou
     {
         for (;;) select( 0, NULL, NULL, NULL, NULL );
     }
-    else
+    else if (timeout->QuadPart < 0)
     {
+        timeout_t when = -timeout->QuadPart, diff;
         LARGE_INTEGER now;
-        timeout_t when, diff;
 
-        if ((when = timeout->QuadPart) < 0)
+        NtQueryPerformanceCounter( &now, NULL );
+        when += now.QuadPart;
+
+        NtYieldExecution();
+
+        for (;;)
         {
-            NtQuerySystemTime( &now );
-            when = now.QuadPart - when;
+            struct timeval tv;
+            NtQueryPerformanceCounter( &now, NULL );
+            diff = (when - now.QuadPart + 9) / 10;
+            if (diff <= 0) break;
+            tv.tv_sec = diff / 1000000;
+            tv.tv_usec = diff % 1000000;
+            if (select( 0, NULL, NULL, NULL, &tv ) != -1) break;
         }
+    }
+    else
+    {
+        timeout_t when = timeout->QuadPart, diff;
+        LARGE_INTEGER now;
 
         /* Note that we yield after establishing the desired timeout, but
            we only care about the result of the yield for zero timeouts */
-- 
2.55.0

